libopae/plugins/xfpga/opae_ioctl.c

```diff
@@ -286,7 +331,262 @@ int opae_fme_port_pr(int fd, uint32_t flags, uint32_t port_id, uint32_t sz,
+{
+	ASSERT_NOT_NULL(info);
+	struct dfl_fpga_port_info pinfo = {.argsz = sizeof(pinfo), .flags = 0};
+	int res = opae_ioctl(fd, DFL_FPGA_PORT_GET_INFO, &pinfo);
+	if (!res) {
+		info->flags = pinfo.flags;
+		info->num_regions = pinfo.num_regions;
+		info->num_umsgs = pinfo.num_umsgs;
+	}
+	return res;
+}
+
+fpga_result dfl_get_port_region_info(int fd, uint32_t index,
+				     opae_port_region_info *info)
+{
+	ASSERT_NOT_NULL(info);
+	struct dfl_fpga_port_region_info rinfo = {
+		.argsz = sizeof(rinfo), .padding = 0, .index = index};
+	int res = opae_ioctl(fd, DFL_FPGA_PORT_GET_REGION_INFO, &rinfo);
+	if (!res) {
+		info->flags = rinfo.flags;
+		info->size = rinfo.size;
+		info->offset = rinfo.offset;
+	}
+	return res;
+}
+
+fpga_result dfl_port_map(int fd, void *addr, uint64_t len, uint64_t *io_addr)
+{
+	int res = 0;
+	/* Set ioctl fpga_port_dma_map struct parameters */
+	struct dfl_fpga_port_dma_map dma_map = {.argsz = sizeof(dma_map),
+						.flags = 0,
+						.user_addr = (__u64)addr,
+						.length = (__u64)len,
+						.iova = 0};
+	ASSERT_NOT_NULL(io_addr);
+	/* Dispatch ioctl command */
+	res = opae_ioctl(fd, DFL_FPGA_PORT_DMA_MAP, &dma_map);
+	if (!res) {
+		*io_addr = dma_map.iova;
+	}
+	return res;
+}
+
+fpga_result dfl_port_unmap(int fd, uint64_t io_addr)
+{
+	/* Set ioctl fpga_port_dma_unmap struct parameters */
+	struct dfl_fpga_port_dma_unmap dma_unmap = {
+		.argsz = sizeof(dma_unmap), .flags = 0, .iova = io_addr};
+
+	/* Dispatch ioctl command */
+	return opae_ioctl(fd, DFL_FPGA_PORT_DMA_UNMAP, &dma_unmap);
+}
+
+
+fpga_result dfl_fme_port_assign(int fd, uint32_t flags, uint32_t port_id)
+{
+	struct dfl_fpga_fme_port_assign assign = {
+		.argsz = sizeof(assign), .flags = 0, .port_id = port_id};
+	if (flags) {
+		OPAE_MSG(
+			"flags currently not supported in DFL_FPGA_FME_PORT_ASSIGN");
+	}
+	return opae_ioctl(fd, DFL_FPGA_FME_PORT_ASSIGN, &assign);
+}
+
+fpga_result dfl_fme_port_release(int fd, uint32_t flags, uint32_t port_id)
+{
+	struct dfl_fpga_fme_port_assign assign = {
+		.argsz = sizeof(assign), .flags = 0, .port_id = port_id};
+	if (flags) {
+		OPAE_MSG(
+			"flags currently not supported in DFL_FPGA_FME_PORT_RELEASE");
+	}
+	return opae_ioctl(fd, DFL_FPGA_FME_PORT_RELEASE, &assign);
+}
+
+fpga_result dfl_fme_port_pr(int fd, uint32_t flags, uint32_t port_id,
+			    uint32_t sz, uint64_t addr, uint64_t *status)
+{
+	struct dfl_fpga_fme_port_pr port_pr = {.argsz = sizeof(port_pr),
+					       .flags = 0,
+					       .port_id = port_id,
+					       .buffer_size = sz,
+					       .buffer_address = addr};
+	int res = FPGA_OK;
+	if (flags) {
+		OPAE_MSG("flags currently not supported in FPGA_FME_PORT_PR");
+	}
+	ASSERT_NOT_NULL(status);
+	res = opae_ioctl(fd, DFL_FPGA_FME_PORT_PR, &port_pr);
+	*status = 0;
+	return res;
+}
+
+fpga_result dfl_fme_port_reset(int fd)
+{
+	return opae_ioctl(fd, DFL_FPGA_PORT_RESET, NULL);
+}
+
+#define MAX_KERNEL_DRIVERS 2
+static ioctl_ops ioctl_table[MAX_KERNEL_DRIVERS] = {
```

Comment:
More precisely, ioctl_table[0] is for patch set 0 of the up-streamed driver, and we anticipate extending the table for future upstreamed patchsets?

Reply:
I think so but only if the changes to future updates are not backwards compatible.
If future patch sets contain new ioctl request codes, the header will be updated and the table will include function pointers to the new codes. If these functions are used on an older patch set, like patch set 0, with the newer codes, we should get an error code that translates to FPGA_NOT_SUPPORTED.